demos/device-shadow-for-aws-iot-embedded-sdk/shadow_demo_helpers.c

```diff
@@ -0,0 +1,916 @@
+
+/**
+ * @brief Connect to MQTT broker with reconnection retries.
+ *
+ * If connection fails, retry is attempted after a timeout.
+ * Timeout value will exponentially increase until maximum
+ * timeout value is reached or the number of attempts are exhausted.
+ *
+ * @param[out] pxNetworkContext The output parameter to return the created network context.
+ *
+ * @return The status of the final connection attempt.
+ */
+static TransportSocketStatus_t prvConnectToServerWithBackoffRetries( NetworkContext_t * pxNetworkContext );
+
+/**
+ * @brief Function to get the free index at which an outgoing publish
+ * can be stored.
+ *
+ * @param[out] pucIndex The output parameter to return the index at which an
+ * outgoing publish message can be stored.
+ *
+ * @return EXIT_FAILURE if no more publishes can be stored;
+ * EXIT_SUCCESS if an index to store the next outgoing publish is obtained.
+ */
+static int prvGetNextFreeIndexForOutgoingPublishes( uint8_t * pucIndex );
+
+/**
+ * @brief Function to clean up an outgoing publish at given index from the
+ * #outgoingPublishPackets array.
+ *
+ * @param[in] ucIndex The index at which a publish message has to be cleaned up.
+ */
+static void vCleanupOutgoingPublishAt( uint8_t ucIndex );
+
+/**
+ * @brief Function to clean up all the outgoing publishes maintained in the
+ * array.
+ */
+static void vCleanupOutgoingPublishes( void );
+
+/**
+ * @brief Function to clean up the publish packet with the given packet id.
+ *
+ * @param[in] usPacketId Packet identifier of the packet to be cleaned up from
+ * the array.
+ */
+static void vCleanupOutgoingPublishWithPacketID( uint16_t usPacketId );
+
+/**
+ * @brief Function to resend the publishes if a session is re-established with
+ * the broker. This function handles the resending of the QoS1 publish packets,
+ * which are maintained locally.
+ *
+ * @param[in] pxMqttContext MQTT context pointer.
+ */
+static int handlePublishResend( MQTTContext_t * pxMqttContext );
+
+/**
+ * @brief The timer query function provided to the MQTT context.
+ *
+ * @return Time in milliseconds.
+ */
+static uint32_t prvGetTimeMs( void );
+
+/*-----------------------------------------------------------*/
+
+static TransportSocketStatus_t prvConnectToServerWithBackoffRetries( NetworkContext_t * pxNetworkContext )
+{
+    TransportSocketStatus_t xNetworkStatus = TRANSPORT_SOCKET_STATUS_SUCCESS;
+    RetryUtilsStatus_t xRetryUtilsStatus = RetryUtilsSuccess;
+    RetryUtilsParams_t xReconnectParams = { 0 };
+    ServerInfo_t xServerInfo = { 0 };
+    SocketsConfig_t xSocketConfig = { 0 };
+
+    /* Initialize information to connect to the MQTT broker. */
+    xServerInfo.pHostName = democonfigMQTT_BROKER_ENDPOINT;
+    xServerInfo.hostNameLength = sizeof( democonfigMQTT_BROKER_ENDPOINT ) - 1U;
+    xServerInfo.port = democonfigMQTT_BROKER_PORT;
+
+    /* Set the Secure Socket configurations. */
+    /* Configure credentials for TLS mutual authenticated session. */
```

Comment:
The TLS mutual authentication session is not a proper noun.

```suggestion
    /* Configure credentials for a TLS mutual authenticated session. */
```

Reply:
I will remove this line because it's the comment from csdk and we don't configure credential here.